store/tikv/region_request.go

```diff
@@ -182,12 +187,183 @@ func (s *RegionRequestSender) SetRPCError(err error) {
+		if !s.region.isValid() {
+			return nil, nil
+		}
+		if s.isExhausted() {
+			s.invalidateRegion()
+			return nil, nil
+		}
+		replica := s.replicas[s.nextReplicaIdx]
+		s.nextReplicaIdx++
+
+		// Limit the max attempts of each replica to prevent endless retry.
+		replica.attempts++
+		if replica.attempts > maxReplicaAttempt {
+			continue
+		}
+
+		storeFailEpoch := atomic.LoadUint32(&replica.store.epoch)
+		if storeFailEpoch != replica.epoch {
+			// TODO(youjiali1995): Is it necessary to invalidate the region?
+			s.invalidateRegion()
+			return nil, nil
+		}
+		addr, err := s.regionCache.getStoreAddr(bo, s.region, replica.store)
+		if err == nil && len(addr) != 0 {
+			return &RPCContext{
+				Region:     s.region.VerID(),
+				Meta:       s.region.meta,
+				Peer:       replica.peer,
+				Store:      replica.store,
+				Addr:       addr,
+				AccessMode: TiKVOnly,
+				TiKVNum:    len(s.replicas),
+			}, nil
+		}
+	}
+}
+
+func (s *replicaSelector) onSendFailure(bo *Backoffer, err error) {
+	metrics.RegionCacheCounterWithSendFail.Inc()
+	replica := s.replicas[s.nextReplicaIdx-1]
+	if replica.store.requestLiveness(bo, s.regionCache) == reachable {
+		s.rewind()
+		return
+	}
+
+	store := replica.store
+	// invalidate regions in store.
+	if atomic.CompareAndSwapUint32(&store.epoch, replica.epoch, replica.epoch+1) {
+		logutil.BgLogger().Info("mark store's regions need be refill", zap.Uint64("id", store.storeID), zap.String("addr", store.addr), zap.Error(err))
+		metrics.RegionCacheCounterWithInvalidateStoreRegionsOK.Inc()
+		// schedule a store addr resolve.
+		store.markNeedCheck(s.regionCache.notifyCheckCh)
+	}
+	// TODO(youjiali1995): It's not necessary, but some tests depend on it and it's not easy to fix.
+	if s.isExhausted() {
+		s.region.scheduleReload()
+	}
+}
+
+// OnSendSuccess updates the leader of the cached region since the replicaSelector
+// is only used for leader request. It's called when the request is sent to the
+// replica successfully.
+func (s *replicaSelector) OnSendSuccess() {
+	// The successful replica is not at the head of replicas which means it's not the
+	// leader in the cached region, so update leader.
+	if s.nextReplicaIdx-1 != 0 {
+		leader := s.replicas[s.nextReplicaIdx-1].peer
+		if !s.regionCache.switchWorkLeaderToPeer(s.region, leader) {
+			panic("the store must exist")
+		}
+	}
+}
+
+func (s *replicaSelector) rewind() {
+	s.nextReplicaIdx--
+}
+
+// updateLeader updates the leader of the cached region.
+// If the leader peer isn't found in the region, the region will be invalidated.
+func (s *replicaSelector) updateLeader(leader *metapb.Peer) {
+	if leader == nil {
+		return
+	}
+	for i, replica := range s.replicas {
+		if isSamePeer(replica.peer, leader) {
+			if i < s.nextReplicaIdx {
+				s.nextReplicaIdx--
+			}
+			// Move the leader replica to the front of candidates and skip the current replica.
+			s.replicas[i], s.replicas[s.nextReplicaIdx] = s.replicas[s.nextReplicaIdx], s.replicas[i]
```

Comment:
Should we clear the attempts of the leader replica? Otherwise it seems we might skip it because exceeding max attemp limit.

Reply:
I intend to do it, otherwise, it may result in infinite retry and if the leader has been tried many times, it's nearly impossible to recover the next time.

Reply:
i see, makes sense

Reply:
Emm, but if we don't believe the leader returned by this replica, should we try this replica again? Then, we should add a `continue` after L343?

Reply:
I'm thinking about it. My concern is the maximum retry count is 10*replica count in such a case.